File: python/_102.py
# BFS using deque
# O(N) / O(N)
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        
        queue = collections.deque([root])
        ans = []
        
        while queue:
            level_size = len(queue)
            cur_level = []
            for _ in range(level_size):
                node = queue.popleft()
                cur_level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            ans.append(cur_level)
            
        return ans


# BFS not using deque
# O(N) / O(N)
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        
        arr, ans = [root], []
        while arr:
            nxt_arr, cur_level = [], []
            for node in arr:
                cur_level.append(node.val)
                if node.left:
                    nxt_arr.append(node.left)
                if node.right:
                    nxt_arr.append(node.right)
            arr = nxt_arr
            ans.append(cur_level)
            
        return ans

File: cpp/_102.cpp
// Solution 1. BFS using queue
// O(N) / O(N)
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            int level_size = q.size();
            vector<int> cur_level;
            for (int i = 0; i < level_size; i++) {
                TreeNode* node = q.front();
                q.pop();
                cur_level.push_back(node->val);
                if (node->left)
                    q.push(node->left);
                if (node->right)
                    q.push(node->right);
            }
            ans.push_back(cur_level);
        }
        
        return ans;
    }
};


// Solution 2. BFS not using queue
// O(N) / O(N)
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        
        vector<vector<int>> ans;
        vector<TreeNode*> arr;
        arr.push_back(root);
        
        while (!arr.empty()) {
            vector<TreeNode*> nxt_arr;
            vector<int> cur_level;
            for (TreeNode* node : arr) {
                cur_level.push_back(node->val);
                if (node->left)
                    nxt_arr.push_back(node->left);
                if (node->right)
                    nxt_arr.push_back(node->right);
            }
            arr = nxt_arr;
            ans.push_back(cur_level);
        }
        
        return ans;
    }
};